### DISCUSSION TITLE
Formatting options

### DISCUSSION THREAD

**Discussion by ohler55** [ohler55/ojg#65] [Ideas]
Let discuss how custom formatting options might be provided without adding a new option parameter to OjG for each variation.

**Comment by ohler55** [comment]
Want to chime in @eugenesvk?

**Comment by eugenesvk** [comment]
Given the constraints you described, the only idea I'd had so far: if adding the option complexity in the output binary is too undesireable for performance reasons etc., would then moving this complexity to the compilation step work?

For example, I know that I'd always want to have **separators vertically** aligned (using the example we discussed [here](https://github.com/ohler55/ojg/issues/35#issuecomment-873613837)), but then from your perspective
- it's too much to add an extra config option `--align-separators`
- making it the default is also not great

**As an alternative**, I could build **OjG** with some `--compile-with-align-separators` flag and this version would have the separators always aligned by default (thus, no slowing down due to having to parse various config options), while the "original" OjG version would have no such alignment built in (but also not be slowed down due to having to parse an extra config option)
This way the final code that ends up in the binary for both versions is "minimal" to what is needed

**Reply by ohler55** [reply]
From experience with Oj for Ruby many people wanted different formatting options that varied from the number of decimals to display on numbers to the use of spaces around the `:` separators. I'd like to avoid that situation by providing a hook that people can provide their own modifications to the output with changing the OjG code.

One thought I had was providing a secondary formatting option on a line by line basis. The line would start out as a what OjG outputs but then the secondary formatter could add spaces, slide the `:` around, colorize, or do what ever they like.

**Reply by eugenesvk** [reply]
well, the best :) features are about vertical alignment, so they require info about all the lines and thus wouldn't benefit from a second line-by-line pass (so the `:` wouldn't move to the correct spot?)
Otherwise seems like an interesting idea for some other options

**Reply by ohler55** [reply]
Although if already vertically aligned the modifier could shuffle characters while maintaining the alignment. I'm not even sure how it would be implemented but it seems like it might work when sketching it out on a napkin.
